lib/ui/platform_dispatcher.dart

```diff
@@ -2605,3 +2662,52 @@ class SemanticsActionEvent {
     );
   }
 }
+
+/// Signature for [PlatformDispatcher.onPlatformFocusChange].
+typedef ViewFocusChangeCallback = void Function(ViewFocusEvent);
+
+/// An event for the engine to communicate view focus changes back to the app.
+final class ViewFocusEvent {
+  /// Creates a [ViewFocusChange].
+  const ViewFocusEvent({
+    required this.viewId,
+    required this.focusState,
+  });
+
+  /// The id of the [FlutterView] that experienced a focus change.
+  final int viewId;
+
+  /// The current [ViewFocusState] of the [FlutterView] with the given [viewId].
+  final ViewFocusState viewFocusState;
+}
+
+/// Represents a given flutter view focus state.
+///
+/// If focus was lost, use [ViewFocusState.unfocused].
+/// If focus was gained without direction, use [ViewFocusState.focused].
+/// If focus was gained in forward direction, use [ViewFocusState.forwardFocused].
+/// If focus was gained in backwards direction, use [ViewFocusState.backwardsFocused].
+///
+/// Valid transitions, within a view, are:
+///
+/// - [ViewFocusState.unfocused] to [ViewFocusState.focused].
+/// - [ViewFocusState.unfocused] to [ViewFocusState.forwardFocused].
+/// - [ViewFocusState.unfocused] to [ViewFocusState.backwardsFocused].
+/// - [ViewFocusState.focused] to [ViewFocusState.unfocused].
+/// - [ViewFocusState.forwardFocused] to [ViewFocusState.unfocused].
+/// - [ViewFocusState.backwardsFocused] to [ViewFocusState.unfocused].
+///
+/// All others are invalid.
+enum ViewFocusState {
```

Comment:
I wonder if this really are two separate enums. One for focused/unfocuses and one for the direction, where in the ViewFocusEvent the direction would be optional/nullable. Not 100% sure, though.

Reply:
This is an interesting idea. It would allow us to expand the space in the future without polluting the enum's value space. For example, today we have "forward", "backward", and "no direction" that apply only to "gain focus" event. However, we may find in the future that we'll want to communicate:

- Lose focus in the backward direction.
- Gain focus and focus on the last-focused widget.
- Gain focus and supply custom data for the app to decide what to focus on.

For extensibility, the focus state can remain an `enum`, but the target of the focus could be a `sealed class`, e.g.:

```dart
enum ViewFocusState {
  focused,
  unfocused,
}

sealed class ViewFocusTarget {}

enum ViewFocusDirectionalTarget implements ViewFocusTarget {
  forward,
  backward,
  lastFocusedWidget,
}

final class ViewFocusCustomTarget implements ViewFocusTarget {
  ViewFocusCustomTarget(this.data);
  
  final Object data;
}

void handleViewFocusRequest(ViewFocusState state, ViewFocusTarget target) {
  switch (target) {
    case ViewFocusDirectionalTarget():
      print('got ViewFocusDirectionalTarget');
      switch (target) {
        case ViewFocusDirectionalTarget.forward:
          print('forward');
        case ViewFocusDirectionalTarget.backward:
          print('backward');
        case ViewFocusDirectionalTarget.lastFocusedWidget:
          print('lastFocusedWidget');
      }
    case ViewFocusCustomTarget():
      print('got ViewFocusCustomTarget with data ${target.data}');
  }
}
```

I suggest we do not speculatively use this structure for the MVP. However, we could subset it so that we could expand it in the future.

Reply:
Seems like a lot more code compared to just adding an entry to an enum. What are the advangates of such proposals? 

Or in a slightly different way: What are the dissadvantages of having just an enum?

Reply:
Done